src/locktower.rs

```diff
@@ -0,0 +1,172 @@
+use std::collections::VecDeque;
+pub struct VoteLock {
+    /// Fork id, same as slot id. There can only be 1 fork per slot, since the empty PoH ticks
+    /// cannot be voted on.  A validator cannot vote on a fork unless that fork is connected to
+    /// a trunk. The starting trunk is the initial `finalized_fork`.
+    fork: u64,
+    /// Lockout is in terms of slots, so the slot height or fork id to unroll this
+    /// vote must be greater than fork + lockout.
+    lockout: u64,
+}
+
+pub enum LockTowerError {
+    VoteLocked,
+    VoteNotConverging,
+}
+
+pub struct LockTower {
+    vote_locks: VecDeque<VoteLock>,
+    /// The maximum size of vote_locks before reward is generated.  When the queue reached max_size,
+    /// a reward is generated and the oldest lock is removed.  That fork is considered finalized.
+    max_size: usize,
+    /// Depth is equal to lockout, since lockout doubles as votes get stacked
+    /// If the requirement is at depth 8 to have converge_amount of 0.667,
+    /// than the lockout will not increase past 256 (2^8) on the fork at depth 8
+    /// without 66.7% total network commitment to that fork.
+    ///
+    /// It is not important for the rest of the network to also have the same lockout on the fork.
+    /// The worst case is that the lockout for this node will double to 512 due to
+    /// rollback for the rest of the network.  So any lockout of network commitment is acceptable,
+    /// but it must cover the required threshold percentage of network stake.
+    converge_depth: usize,
+    converge_amount: f64,
+    /// The last finalized fork.  This value is initialized at the start with some pre-defined
+    /// network value, such as fork 0.
+    finalized_fork: u64,
+}
+
+/// This should be implemented by the Bank
+pub trait ForkTree {
+    /// Return true if the leaf is a descendant of the trunk.
+    fn is_trunk_of(&self, trunk: u64, leaf: u64) -> bool;
+    /// Return what percentage of the network has agreed to the fork or a descendant fork.
+    fn convergance(&self, fork: u64) -> f64;
+}
+
+impl LockTower {
+    pub fn new(
+        max_size: usize,
+        converge_depth: usize,
+        converge_amount: f64,
+        finalized_fork: u64,
+    ) -> Self {
+        LockTower {
+            vote_locks: VecDeque::new(),
+            max_size,
+            converge_depth,
+            converge_amount,
+            finalized_fork,
+        }
+    }
+
+    /// * vote - fork id to vote on
+    /// Returns true if this vote is unlocked.
+    pub fn is_unlocked<T: ForkTree>(&self, vote: u64, tree: &T) -> bool {
+        let ix = self.num_cleared_votes(vote);
+        self.vote_locks
+            .get(ix)
+            .map(|lockout| tree.is_trunk_of(lockout.fork, vote))
+            .unwrap_or(tree.is_trunk_of(self.finalized_fork, vote))
```

Comment:
use `unwrap_or_else(|| ...)`. This will call `is_trunk_of` regardless to pass `unwrap_or` its result.